### PR TITLE
Fix false call setup warning when property index is a non-virtual member

### LINKED ISSUE #1072: Non-Virtual Setup Analyzer gives false positive when complex argument constraint used in indexer
Found in [the Nancy codebase](https://github.com/NancyFx/Nancy/blob/be4f8d42076e4e568a3742715437868e6c7d05af/test/Nancy.ViewEngines.Razor.Tests/TextResourceFinderFixture.cs#L68).
Can be reproduced by

```c#
namespace TheNamespace
{
    using FakeItEasy;

    public interface IHaveAnIndexer { string this[int context] { get; } }

    public class TheClass
    {
        public void Test()
        {
            var fake = A.Fake<IHaveAnIndexer>();
            A.CallTo(() => fake[A<int>._]).MustHaveHappened();
        }
    }
}
```

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #1072.

### CODE REVIEW

**Comment by thomaslevesque** [comment]
Uh... apparently the change I suggested breaks things for VB.NET. Looking into it.

**Comment by thomaslevesque** [comment]
OK, I have something that seems to work: https://github.com/thomaslevesque/FakeItEasy/commits/indexer-analyzer

**Comment by blairconrad** [comment]
@thomaslevesque, your comments have merit, as always. I think maybe this should be your PR now, but I've merged your commits into mine because it was the easiest way for me to make comments. Comments forthcoming.

File: src/FakeItEasy.Analyzer.Shared/NonVirtualSetupAnalyzer.cs [hunk 1]
```diff
@@ -128,9 +160,55 @@ private static bool IsMethod(SymbolInfo symbolInfo)
...
+
+            return false;
+        }
+
         private static bool IsInterfaceMember(SymbolInfo symbolInfo)
         {
             return symbolInfo.Symbol?.ContainingType?.TypeKind == TypeKind.Interface;
         }
+
+        private static InvocationExpressionSyntax FindInvocationInHierarchy(SyntaxNode node, IEnumerable<Ancestor> ancestors)
```
Comment (blairconrad) [inline src/FakeItEasy.Analyzer.Shared/NonVirtualSetupAnalyzer.cs]: `ancestors` sounds to me like these are the actual ancestors, not the ones we require (or expect). Perhaps a name change here and for `NodeAncestors`?

Now that I notice it, we _could_ opt not to pass in `ancestors` and just use `NodeAncestors` directly.

File: tests/FakeItEasy.Analyzer.VisualBasic.Tests/NonOverridableSetupAnalyzerTests.cs [hunk 2]
```diff
@@ -259,6 +259,83 @@ End Namespace
...
+            A.CallTo(Function() fake(A(Of Integer).Ignored)).MustNotHaveHappened()
+        End Sub
+    End Class
+End Namespace
+";
+            this.VerifyVisualBasicDiagnostic(Test);
+        }
+
+        [Fact]
+        public void Diagnostic_Should_Not_Be_Triggered_For_Non_Overridable_Indexer()
```
Comment (blairconrad) [inline tests/FakeItEasy.Analyzer.VisualBasic.Tests/NonOverridableSetupAnalyzerTests.cs]: "Overridable_Indexer"

File: src/FakeItEasy.Analyzer.Shared/NonVirtualSetupAnalyzer.cs (hunk 1 above)
Comment (thomaslevesque) [inline src/FakeItEasy.Analyzer.Shared/NonVirtualSetupAnalyzer.cs]: > ancestors sounds to me like these are the actual ancestors, not the ones we require (or expect). Perhaps a name change here and for NodeAncestors?

What do you have in mind? Something like `expectedAncestors` or `requiredAncestors`?

> Now that I notice it, we could opt not to pass in ancestors and just use NodeAncestors directly.

Good point.

File: tests/FakeItEasy.Analyzer.VisualBasic.Tests/NonOverridableSetupAnalyzerTests.cs (hunk 2 above)
Comment (thomaslevesque) [inline tests/FakeItEasy.Analyzer.VisualBasic.Tests/NonOverridableSetupAnalyzerTests.cs]: Oops

File: src/FakeItEasy.Analyzer.Shared/NonVirtualSetupAnalyzer.cs (hunk 1 above)
Comment (blairconrad) [inline src/FakeItEasy.Analyzer.Shared/NonVirtualSetupAnalyzer.cs]: Perhaps `expectedHierarchy`?

Comment (thomaslevesque) [inline src/FakeItEasy.Analyzer.Shared/NonVirtualSetupAnalyzer.cs]: Sounds good

**Comment by thomaslevesque** [comment]
I pushed the changes to your branch

**Comment by blairconrad** [comment]
Squashed!

**Pull request merged**